pep-0701.rst

```diff
@@ -0,0 +1,366 @@
+    >>> f"{joined}"
+    'hello\nworld'
+
+   It only feels natural to allow backslashes in the expression part now that the new PEG parser
+   can easily support it.
+
+    >>> a = ["hello", "world"]
+    >>> f"{'\n'.join(a)}"
+    'hello\nworld'
+
+#. Before the changes proposed in this document, there was no explicit limit in
+   how f-strings can be nested, but the fact that string quotes cannot be reused
+   inside the expression component of f-strings made it impossible to nest
+   f-strings arbitrarily. In fact, this is the most nested-fstring that can be
+   written::
+
+    >>> f"""{f'''{f'{f"{1+1}"}'}'''}"""
+    '2'
+
+   As this PEP allows placing **any** valid Python expression inside the
+   expression component of the f-strings, it is now possible to reuse quotes and
+   therefore is possible to nest f-strings arbitrarily::
+
+    >>> f"{f"{f"{f"{f"{f"{1+1}"}"}"}"}"}"
+    '2'
+   
+   Although this is just a consequence of allowing arbitrary expressions, the
+   authors of this PEP do not believe that this is a fundamental benefit and we
+   have decided that the language specification will not explicitly mandate that
+   this nesting can be arbitrary. This is because allowing arbitrarily-deep
+   nesting imposes a lot of extra complexity to the lexer implementation
+   (particularly as lexer/parser pipelines need to allow "untokenizing" to
+   support the 'f-string debugging expressions' and this is especially taxing when
+   arbitrary nesting is allowed). Implementations are therefore free to impose a
+   limit on the nesting depth if they need to. Note that this is not an uncommon
+   situation, as the CPython implementation already imposes several limits all
+   over the place, including a limit on the nesting depth of parentheses and
+   brackets, a limit on the nesting of the blocks, a limit in the number of
+   branches in ``if`` statements, a limit on the number of expressions in
+   star-unpacking, etc.
+
+Specification
+=============
+
+The formal proposed PEG grammar specification for f-strings is (see :pep:`617`
+for details on the syntax):
+
+.. code-block:: peg
+
+    fstring
+        | FSTRING_START fstring_middle* FSTRING_END
+    fstring_middle
+        | fstring_replacement_field
+        | FSTRING_MIDDLE
+    fstring_replacement_field
+        | '{' (yield_expr | star_expressions) "="? [ "!" NAME ] [ ':' fstring_format_spec* ] '}'
+    fstring_format_spec:
+        | FSTRING_MIDDLE
+        | fstring_replacement_field
+
+This PEP leaves up to the implementation the level of f-string nesting allowed.
+This means that limiting nesting is **not part of the language specification**
+but also the language specification **doesn't mandate arbitrary nesting**. 
+
+Three new tokens are introduced:
+
+* ``FSTRING_START``: This token includes f-string character (``f``/``F``) and the open quote(s).
+* ``FSTRING_MIDDLE``: This token includes everything between two expression braces (``}`` and ``{``).
```

Comment:
As well as the text between the opening quote and the first `{`. Would be good to mention this explicitly:
```suggestion
* ``FSTRING_MIDDLE``: This token includes the text between the opening quote
  and the first expression brace (``{``) and the text between two expression braces (``}`` and ``{``).
```
It was a little unexpected to me that the piece between the last brace and the closing quote is a different token, any reason for that?

Reply:
It simplifies the implementation of the lexer, at least in CPython, but indeed is a bit weird. Is also weird that the last quite is not included.